go-controller/pkg/util/dpu_annotations.go

```diff
@@ -78,16 +78,30 @@ func UnmarshalPodDPUConnDetailsAllNetworks(annotations map[string]string) (map[s
 	return podDcds, nil
 }
 
-// MarshalPodDPUConnDetails adds the pod's connection details of the specified network to the corresponding pod annotation.
-func MarshalPodDPUConnDetails(annotations map[string]string, dcd *DPUConnectionDetails, netName string) (map[string]string, error) {
+// MarshalPodDPUConnDetails adds the pod's connection details of the specified NAD to the corresponding pod annotation;
+// if dcd is nil, delete the pod's connection details of the specified NAD
+func MarshalPodDPUConnDetails(annotations map[string]string, dcd *DPUConnectionDetails, nadName string) (map[string]string, error) {
 	if annotations == nil {
 		annotations = make(map[string]string)
 	}
 	podDcds, err := UnmarshalPodDPUConnDetailsAllNetworks(annotations)
 	if err != nil {
 		return nil, err
 	}
-	podDcds[netName] = *dcd
+	dc, ok := podDcds[nadName]
+	if dcd != nil {
+		if ok && dc == *dcd {
+			return nil, newAnnotationAlreadySetError("OVN pod %s annotation for NAD %s already exists in %v",
+				DPUConnectionDetailsAnnot, nadName, annotations)
+		}
+		podDcds[nadName] = *dcd
+	} else {
+		if !ok {
+			return nil, newAnnotationAlreadySetError("OVN pod %s annotation for NAD %s already removed",
+				DPUConnectionDetailsAnnot, nadName)
+		}
+		delete(podDcds, nadName)
```

Comment:
This function is called `MarshalPodDPUConnDetails`, but we are deleting the Pod's connection details, which is confusing. When will we hit this case? Shouldn't this be an unexpected error? Especially since we have `UnmarshalPodDPUConnDetails`.

Reply:
dpu connection details annotation for a NAD could be deleted as a result of cni cmdDel request. See https://github.com/cathy-zhou/ovn-kubernetes/blob/dpuCNIFix/go-controller/pkg/cni/cni.go#L192

Reply:
Got it thanks!